packages/teko/src/Teko_Utilities.cpp

```diff
@@ -2964,4 +2965,40 @@ std::string formatBlockName(const std::string &prefix, int i, int j, int nrow) {
   return ss.str();
 }
 
+void writeMatrix(const std::string &filename, const Teko::LinearOp &op) {
+  using Teuchos::RCP;
+  using Teuchos::rcp_dynamic_cast;
+#ifdef TEKO_HAVE_EPETRA
+  const RCP<const Thyra::EpetraLinearOp> eOp = rcp_dynamic_cast<const Thyra::EpetraLinearOp>(op);
+#endif
+
+  if (Teko::TpetraHelpers::isTpetraLinearOp(op)) {
+    const RCP<const Thyra::TpetraLinearOp<ST, LO, GO, NT>> tOp =
+        rcp_dynamic_cast<const Thyra::TpetraLinearOp<ST, LO, GO, NT>>(op);
+    const RCP<const Tpetra::CrsMatrix<ST, LO, GO, NT>> crsOp =
+        rcp_dynamic_cast<const Tpetra::CrsMatrix<ST, LO, GO, NT>>(tOp->getConstTpetraOperator(),
+                                                                  true);
+    using Writer = Tpetra::MatrixMarket::Writer<Tpetra::CrsMatrix<ST, LO, GO, NT>>;
+    Writer::writeMapFile(("rowmap_" + filename).c_str(), *(crsOp->getRowMap()));
+    Writer::writeMapFile(("colmap_" + filename).c_str(), *(crsOp->getColMap()));
+    Writer::writeMapFile(("domainmap_" + filename).c_str(), *(crsOp->getDomainMap()));
+    Writer::writeMapFile(("rangemap_" + filename).c_str(), *(crsOp->getRangeMap()));
+    Writer::writeSparseFile(filename.c_str(), crsOp);
+  }
+#ifdef TEKO_HAVE_EPETRA
+  else if (eOp != Teuchos::null) {
+    const RCP<const Epetra_CrsMatrix> crsOp =
+        rcp_dynamic_cast<const Epetra_CrsMatrix>(eOp->epetra_op(), true);
+    EpetraExt::BlockMapToMatrixMarketFile(("rowmap_" + s).c_str(), crsOp->RowMap());
+    EpetraExt::BlockMapToMatrixMarketFile(("colmap_" + s).c_str(), crsOp->ColMap());
+    EpetraExt::BlockMapToMatrixMarketFile(("domainmap_" + s).c_str(), crsOp->DomainMap());
+    EpetraExt::BlockMapToMatrixMarketFile(("rangemap_" + s).c_str(), crsOp->RangeMap());
+    EpetraExt::RowMatrixToMatrixMarketFile(s.c_str(), *crsOp);
+  }
+#endif
+  else {
+    TEUCHOS_ASSERT(false);
```

Comment:
Might be worth adding some helpful error message here? A user could inadvertently call this with a `Teko::BlockedLinearOp`, right?